### ISSUE TITLE
Legacy random generator support

### ISSUE THREAD

**Issue by qiyunzhu** [scikit-bio/scikit-bio#2124]
An important question about random generator support remains in scikit-bio. I am pinging @wasade because this question also impacts biom-format, and I would like also to seek @mortonjt and @mataton 's opinion.

## Current situation

Since the new NumPy generator ([`Generator`](https://numpy.org/devdocs/reference/random/generator.html), see [NEP 19](https://numpy.org/neps/nep-0019-rng-policy.html)) was adopted in scikit-bio (#1889) and biom-format (biocore/biom-format#935), users have been able to make results reproducible by specifying a random seed or random generator instance:

```python
res = subsample_counts(counts, n, seed=42)
```

Or:

```python
rng = np.random.default_rng(42)
res = subsample_counts(counts, n, seed=rng)
```

This was a nice improvement. However, it meanwhile dropped the support for legacy random generator ([`RandomState`](https://numpy.org/doc/stable/reference/random/legacy.html)). The impact is that **legacy code** that starts with `np.random.seed()`, which many users might have done, **loses reproducibility**. For example, the following code will produce different results across runs:

```python
np.random.seed(42)
res = subsample_counts(counts, n)
```

## Solution 1

To make the result of the above code reproducible, one can add the following code to the [`get_rng`](https://github.com/scikit-bio/scikit-bio/blob/e6417514c47fce9e140e3017d7c9c7f7e4ca1728/skbio/util/_misc.py#L211) helper function. Tested and working.

```python
if seed is None:
    seed = np.random.randint(1 << 31)
    return np.random.default_rng(seed)
```

However, this solution comes with a shortage: If one runs it using scikit-bio 0.5.x (prior to the adoption of the new random generator) or using 0.6.x, the results will be different. That's because the new and legacy random generators use different algorithms for random number generation. There is no way to make them produce the same output.

## Solution 2

Popular packages such as [SciPy](https://github.com/scipy/scipy/blob/92d2a8592782ee19a1161d0bf3fc2241ba78bb63/scipy/_lib/_util.py#L245), [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/8a2d5ffa0512873a75da608eb14832253979ec44/sklearn/utils/validation.py#L1411) and [Pandas](https://github.com/pandas-dev/pandas/blob/d9cdd2ee5a58015ef6f4d15c7226110c9aab8140/pandas/core/common.py#L433) (click links to see their code) support both new and legacy random generators. They default to the legacy random generator (perhaps for backward compatibility), but they support the new random generator.

I implemented a similar solution in #2123 . Tested and working. The core code is like:

```python
if seed is None:
    return np.random.mtrand._rand
```

However, this solution also has an important limitation. The developer has to consider the different API of the new and legacy random generators while coding. For example, the new generator has the method `integers`, which corresponds to the method `randint` of the legacy generator. It is easy for a developer to omit this and write code that passes tests but may fail in a user's computer.

The current scikit-bio and biom-format codebase has several places where the methods of the new generator is not compatible with the legacy one. Here is a [list](https://github.com/scikit-bio/scikit-bio/actions/runs/10713439416/job/29705469539?pr=2123) of them. One of such incompatibilities stems from biom-format, specifically [here](https://github.com/biocore/biom-format/blob/0692d2f3d6dfdad9d653ca21f46a50235d10981d/biom/_subsample.pyx#L100):

```python
permuted = rng.choice(counts_sum, n, replace=False, shuffle=False)
```

To adopt this solution, the parameter `shuffle=False` needs to be removed if `rng` is a legacy generator. For example:

```python
kwargs = {} if isinstance(rng, np.random.RandomState) else {'shuffle': False}
permuted = rng.choice(counts_sum, n, replace=False, **kwargs)
```

***

Note that the random generator mechanism will impact core functions like `pcoa`, `permanova` and `mantel`. They haven't adopted the new generator yet in scikit-bio, therefore results haven't changed. But once they do, they will be impacted by our decision.

***

Which solution should we adopt? The current situation (no change), solution 1, solution 2, or a different one?

**Comment by qiyunzhu** [comment]
Closing as it has been resolved by adopting solution 1.
